include/range/v3/view/cycle.hpp

```diff
@@ -0,0 +1,259 @@
+                    : rng_(&rng), it_(ranges::begin(rng.rng_)), pos_(0)
+                {
+                }
+
+                reference current() const
+                {
+                    return *it_;
+                }
+
+                void next()
+                {
+                    ++pos_;
+                    const auto end = ranges::end(rng_->rng_);
+                    if (it_ != end)
+                    {
+                        ++it_;
+                    }
+                    if (it_ == end)
+                    {
+                        it_ = begin(rng_->rng_);
+                    }
+                }
+
+                bool equal(cursor const &pos) const
+                {
+                    return rng_ == pos.rng_ && it_ == pos.it_ && pos_ == pos.pos_;
+                }
+
+                CONCEPT_REQUIRES(BidirectionalRange<Rng>{})
+                void prev()
+                {
+                    --pos_;
+                    if (it_ == begin(rng_->rng_))
+                    {
+                        it_ = end(rng_->rng_);
+                    }
+                    --it_;
+                }
+
+                CONCEPT_REQUIRES(RandomAccessRange<Rng>())
+                void advance(difference_type n)
+                {
+                    if (n == difference_type{0})
+                    {
+                        return;
+                    }
+
+                    range_difference_t<Rng> distance = ranges::distance(rng_->rng_);
+                    if (distance == difference_type{0})
+                    {
+                        return;
+                    }
+
+                    pos_ += n;
+                    const auto begin = ranges::begin(rng_->rng_);
+                    const auto end = ranges::end(rng_->rng_);
+
+                    const difference_type pos = it_ - begin;
+                    const auto rel_distance = n >= 0 ? distance - pos : pos;
+
+                    if (n >= 0)
+                    {
+                        if (n < rel_distance)
+                        {
+                            it_ = begin + pos + n;
+                        }
+                        else
+                        {
+                            n -= rel_distance;
+                            it_ = begin + n % distance;
+                        }
+                    }
+                    else
+                    {
+                        n *= -1;
+                        if (n <= rel_distance)
+                        {
+                            it_ = begin + pos - n;
+                        }
+                        else
+                        {
+                            n -= rel_distance;
+                            it_ = begin + distance - n % distance;
+                            if (it_ == end)
+                            {
+                                it_ = begin;
+                            }
+                        }
+                    }
+                }
```

Comment:
> One further examination, I think the reason `take` doesn't work is because your view is container-like: the type of the cursor (and hence iterator) differs based on the constness of the view, which is almost certainly incorrect. views have reference semantics - there should not be a non-const overload of `begin_cursor`/`end_cursor`.

Actually there's nothing fundamentally wrong with a view that uses a top-level const this way. It's a common request, despite the fact that I believe it's always wrong. Most of the views are implemented to be sensitive to const *if the underlying view is also sensitive to const*. None of the views provided by range-v3 are.

Why doesn't `take` work on a `cycle` view?